src/librbd/internal.cc

```diff
@@ -1524,23 +1526,57 @@ int validate_pool(IoCtx &io_ctx, CephContext *cct) {
 	return;
       }
 
-      Context *ctx = new C_CopyWrite(m_throttle, m_bl);
-      auto comp = io::AioCompletion::create(ctx);
-
-      // coordinate through AIO WQ to ensure lock is acquired if needed
-      m_dest->io_work_queue->aio_write(comp, m_offset, m_bl->length(),
-                                       std::move(*m_bl),
-                                       LIBRADOS_OP_FLAG_FADVISE_DONTNEED);
+      if (!m_sparse_size) {
+	m_sparse_size = (1 << m_dest->order);
+      }
+        
+      auto *end_op_ctx = new FunctionContext([this](int r) {
+	this->m_throttle->end_op(r);
+      });
+      C_GatherBuilder gather(m_dest->cct, end_op_ctx);
+
+      bufferptr m_ptr(m_bl->length());
+      m_bl->rebuild(m_ptr);
+      size_t write_offset = 0;
+      size_t write_length = 0;
+      size_t offset = 0;
+      size_t length = m_bl->length();
+      while (offset < length) {
+	if (util::calc_sparse_extent(m_ptr,
+				     m_sparse_size,
+				     length,
+				     &write_offset,
+				     &write_length,
+				     &offset)) {
+	  bufferptr write_ptr(m_ptr, write_offset, write_length);
+	  bufferlist *write_bl = new bufferlist();
+	  write_bl->push_back(write_ptr);
+	  Context *ctx = new C_CopyWrite(write_bl, gather.new_sub());
+	  auto comp = io::AioCompletion::create(ctx);
+
+	  // coordinate through AIO WQ to ensure lock is acquired if needed
+	  m_dest->io_work_queue->aio_write(comp, m_offset + write_offset,
+					   write_length,
+					   std::move(*write_bl),
+					   LIBRADOS_OP_FLAG_FADVISE_DONTNEED);
+	  write_offset = offset;
+	  write_length = 0;
+	}
+      }
+      delete m_bl;
+      assert(gather.has_subs());
```

Comment:
yeah, but if 100% sparse， 
it will return at
      if (m_bl->is_zero()) {
        delete m_bl;
        m_throttle->end_op(r);
        return;
      }

So I make this assertion here